flox/core.py

```diff
@@ -1550,25 +1603,53 @@ def _factorize_multiple(by, expected_groups, any_by_dask, reindex):
             *by_,
             chunks=tuple(chunks.values()),
             meta=np.array((), dtype=np.int64),
-            **kwargs,
-        )
-        found_groups = tuple(
-            None if is_duck_dask_array(b) else pd.unique(b.reshape(-1)) for b in by
-        )
-        grp_shape = tuple(
-            len(e) if e is not None else len(f) for e, f in zip(expected_groups, found_groups)
+            expected_groups=expected_groups,
+            axes=(),  # always (), we offset later if necessary.
+            fastpath=True,
+            reindex=reindex,
         )
+
+        fg, gs = [], []
+        for b, e in zip(by, expected_groups):
+            if e is None:
+                if is_duck_dask_array(b):
+                    raise ValueError(
+                        "Please provide expected_groups when grouping by a dask array."
+                    )
+
+                f = pd.unique(b.reshape(-1))
+            else:
+                f = e.to_numpy()
+
+            fg.append(f)
+            gs.append(len(f))
+
+        found_groups = tuple(fg)
+        grp_shape = tuple(gs)
+
+        # found_groups = tuple(
+        #     None if is_duck_dask_array(b) else pd.unique(b.reshape(-1)) for b in by
+        # )
+        # grp_shape = tuple(
+        #     len(e) if e is not None else len(f) for e, f in zip(expected_groups, found_groups)
+        # )
     else:
-        group_idx, found_groups, grp_shape = factorize_(by, **kwargs)
+        group_idx, found_groups, grp_shape, ngroups, size, props = factorize_(
+            by,
+            expected_groups=expected_groups,
+            axes=(),  # always (), we offset later if necessary.
+            fastpath=True,
+            reindex=reindex,
+        )
 
-    final_groups = tuple(
-        found if expect is None else expect.to_numpy()
-        for found, expect in zip(found_groups, expected_groups)
-    )
+    # final_groups = tuple(
+    #     found if expect is None else expect.to_numpy()
+    #     for found, expect in zip(found_groups, expected_groups)
+    # )
```

Comment:
As I understand it `found_groups` from `factorize_` is just a better version of `expected_groups`. 
So I don't see why `expected_groups` in `final_groups` is prioritized. 

Seems to pass the tests skipping this loop at least.

Reply:
It's because `set(expected_groups) < set(found_groups)`  possibly.

Reply:
Sounds like we need a extra test.